File: system-test/typescript/src/testkit/dsl/core/usecase/usecases/BrowseCouponsVerification.ts
import { expect } from '@playwright/test';
import type {
  BrowseCouponsResponse,
  BrowseCouponItem,
} from '../../../../driver/port/dtos/BrowseCouponsResponse.js';
import { ResponseVerification } from '../../shared/response-verification.js';
import type { UseCaseContext } from '../../shared/use-case-context.js';

export class BrowseCouponsVerification extends ResponseVerification<BrowseCouponsResponse> {
  constructor(response: BrowseCouponsResponse, context: UseCaseContext) {
    super(response, context);
  }

  hasCouponWithCode(couponCodeAlias: string): this {
    this.findCouponByCode(couponCodeAlias);
    return this;
  }

  couponHasDiscountRate(couponCodeAlias: string, expectedDiscountRate: number): this {
    const coupon = this.findCouponByCode(couponCodeAlias);
    expect(coupon.discountRate).toBe(expectedDiscountRate);
    return this;
  }

  couponHasValidFrom(couponCodeAlias: string, expectedValidFrom: string): this {
    const coupon = this.findCouponByCode(couponCodeAlias);
    expect(coupon.validFrom).toBe(expectedValidFrom);
    return this;
  }

  couponHasValidTo(couponCodeAlias: string, expectedValidTo: string): this {
    const coupon = this.findCouponByCode(couponCodeAlias);
    expect(coupon.validTo).toBe(expectedValidTo);
    return this;
  }

  couponHasUsageLimit(couponCodeAlias: string, expectedUsageLimit: number): this {
    const coupon = this.findCouponByCode(couponCodeAlias);
    expect(coupon.usageLimit).toBe(expectedUsageLimit);
    return this;
  }

  couponHasUsedCount(couponCodeAlias: string, expectedUsedCount: number): this {
    const coupon = this.findCouponByCode(couponCodeAlias);
    expect(coupon.usedCount).toBe(expectedUsedCount);
    return this;
  }

  private findCouponByCode(couponCodeAlias: string): BrowseCouponItem {
    expect(couponCodeAlias).not.toBeNull();
    expect(this.getResponse()).not.toBeNull();
    expect(this.getResponse().coupons).not.toBeNull();

    const couponCode = this.getContext().getParamValue(couponCodeAlias);

    const coupon = this.getResponse().coupons.find((c) => c.code === couponCode);
    if (!coupon) {
      const available = this.getResponse().coupons.map((c) => c.code);
      throw new Error(
        `Coupon with code '${couponCode}' not found. Available coupons: ${JSON.stringify(available)}`,
      );
    }
    return coupon;
  }
}

File: system/multitier/backend-java/src/testSupport/java/com/mycompany/myshop/backend/support/core/usecase/usecases/BrowseCouponsVerification.java
package com.mycompany.myshop.backend.support.core.usecase.usecases;

import static org.assertj.core.api.Assertions.assertThat;

import com.mycompany.myshop.backend.core.dtos.BrowseCouponsResponse;
import com.mycompany.myshop.backend.support.core.shared.ResponseVerification;

public class BrowseCouponsVerification extends ResponseVerification<BrowseCouponsResponse> {

    public BrowseCouponsVerification(BrowseCouponsResponse response) {
        super(response);
    }

    public BrowseCouponsVerification hasCouponWithCode(String couponCode) {
        findCouponByCode(couponCode);
        return this;
    }

    public BrowseCouponsVerification couponHasDiscountRate(
            String couponCode, String expectedDiscountRate) {
        assertThat(findCouponByCode(couponCode).getDiscountRate())
            .as("discount rate of coupon '%s'", couponCode)
            .isEqualByComparingTo(expectedDiscountRate);
        return this;
    }

    public BrowseCouponsVerification couponHasUsageLimit(String couponCode, int expectedUsageLimit) {
        assertThat(findCouponByCode(couponCode).getUsageLimit())
            .as("usage limit of coupon '%s'", couponCode)
            .isEqualTo(expectedUsageLimit);
        return this;
    }

    public BrowseCouponsVerification couponHasUsedCount(String couponCode, int expectedUsedCount) {
        assertThat(findCouponByCode(couponCode).getUsedCount())
            .as("used count of coupon '%s'", couponCode)
            .isEqualTo(expectedUsedCount);
        return this;
    }

    private BrowseCouponsResponse.BrowseCouponsItemResponse findCouponByCode(String couponCode) {
        assertThat(getResponse().getCoupons()).as("coupons").isNotNull();

        return getResponse().getCoupons().stream()
            .filter(coupon -> couponCode.equals(coupon.getCode()))
            .findFirst()
            .orElseThrow(() -> new AssertionError(String.format(
                "Coupon with code '%s' not found. Available coupons: %s",
                couponCode,
                getResponse().getCoupons().stream()
                    .map(BrowseCouponsResponse.BrowseCouponsItemResponse::getCode)
                    .toList())));
    }
}
